### PR TITLE
Array sum

### PR DESCRIPTION
**Pull request by cd1m0**
This PR includes several things:

1. I forgot to actually merge array unchecked sum functions on fri (brain dead moment).
2. As part of the array unchecked sum functions, we also refactored InstrumentationContext. Specifically, there were several patterns of caches that take some complex arguments, generate a string based key from them, and then store the cached result in a map using the generated key. To reduce duplication we introduced the new `StructMap` and `FactoryMap` classes that contain the logic for building an arbitrary structural-equality based map. 
3. As part of the array unchecked sum functions changeset we also moved a bunch of common instrumentation functions in `instrumenter/utils.ts`
4. Fix a bug in array sum functions where the body was not wrapped in an unchecked block
5. Fix a bug in custom map setter generation, where they were not using the `addKey` helper

### CODE REVIEW

**Review by blitz-1306 — COMMENTED** [review]
Great cleanup! Few small suggestions and good to go.

File: src/instrumenter/instrument.ts [hunk 1]
```diff
@@ -1184,3 +1186,87 @@ export function getOrAddConstructor(
...
     return emptyConstructor;
 }
+
+/**
+ * Given an array type arrT (actually, poitner to array type) and a container, build a function that computes the
+ * sum over an array of type `arrT` and add it to `container`
+ * @param ctx
+ * @param container
+ * @param arrT
+ * @returns
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/instrument.ts]: Fixed in 411ba7e.

File: src/instrumenter/instrumentation_context.ts [hunk 2]
```diff
@@ -138,24 +283,49 @@ export class InstrumentationContext {
...
-
-    private interposingLibraryMap = new Map<string, ContractDefinition>();
-
     public readonly varInterposingQueue: Array<[VariableDeclaration, AbsDatastructurePath]>;
 
     private unitsNeedingUtils = new Set<SourceUnit>();
     private _originalContents: Map<SourceUnit, string>;
     private _aliasedStateVars: Map<VariableDeclaration, ASTNode>;
-    private customMapLibrary = new Map<string, CustomMapLibraryMD>();
+    private _arrSumLibrary?: ContractDefinition;
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/instrumentation_context.ts:296]: Fixed in 411ba7e.

File: src/instrumenter/utils.ts [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            : single(struct.vMembers.filter((field) => field.name === idxArg));
+    return factory.makeMemberAccess("<missing>", base, field.name, field.id);
+}
+
+export function addStructField(
+    factory: ASTNodeFactory,
+    name: string,
+    typ: TypeNode,
+    struct: StructDefinition
+): void {
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/utils.ts]: Good call. Fixed in 411ba7e.

File: src/instrumenter/utils.ts [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        return res;
+    }
+}
+
+/**
+ * Helper function to generate valid function/variable/contract names based on
+ * types.
+ *
+ * @param typ
+ * @returns
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/utils.ts]: Fixed in 411ba7e.

**Review by blitz-1306 — APPROVED** [review]
Cool

**Pull request merged**